### PR TITLE
PLT-44 Inital support for multi-tab multi-team login

### PR DESCRIPTION
**Pull request by coreyhulen**

### CODE REVIEW

File: api/user.go [hunk 1]
```diff
@@ -428,18 +428,9 @@ func Login(c *Context, w http.ResponseWriter, r *http.Request, user *model.User,
...
-		Path:     "/",
-		MaxAge:   maxAge,
-		HttpOnly: true,
-	}
-
-	http.SetCookie(w, sessionCookie)
 
 	multiToken := ""
-	if originalMultiSessionCookie, err := r.Cookie(model.MULTI_SESSION_TOKEN); err == nil {
+	if originalMultiSessionCookie, err := r.Cookie(model.SESSION_COOKIE_TOKEN); err == nil {
```
Comment (coreyhulen) [inline api/user.go]: refactor to common method in context

File: model/client.go [hunk 2]
```diff
@@ -293,7 +295,7 @@ func (c *Client) login(m map[string]string) (*Result, *AppError) {
 	} else {
 		c.AuthToken = r.Header.Get(HEADER_TOKEN)
 		c.AuthType = HEADER_BEARER
-		sessionToken := getCookie(SESSION_TOKEN, r)
+		sessionToken := getCookie(SESSION_COOKIE_TOKEN, r)
```
Comment (coreyhulen) [inline model/client.go]: can be refactored to common method in context

File: web/react/stores/browser_store.jsx [hunk 3]
```diff
@@ -17,32 +17,50 @@ class BrowserStoreClass {
...
         try {
-            localStorage.setItem(name, JSON.stringify(value));
+            if (this.isLocalStorageSupported()) {
+                localStorage.setItem(name, JSON.stringify(value));
+            } else {
+                sessionStorage.setItem(name, JSON.stringify(value));
+            }
         } catch (err) {
             console.log('An error occurred while setting local storage, clearing all props'); //eslint-disable-line no-console
             localStorage.clear();
```
Comment (coreyhulen) [inline web/react/stores/browser_store.jsx]: should be session storage as well

File: web/react/stores/browser_store.jsx [hunk 4]
```diff
@@ -17,32 +17,50 @@ class BrowserStoreClass {
...
     removeItem(name) {
-        localStorage.removeItem(getPrefix() + name);
+        sessionStorage.removeItem(getPrefix() + name);
     }
 
     setGlobalItem(name, value) {
         try {
-            localStorage.setItem(name, JSON.stringify(value));
+            if (this.isLocalStorageSupported()) {
+                localStorage.setItem(name, JSON.stringify(value));
```
Comment (coreyhulen) [inline web/react/stores/browser_store.jsx]: this should have getPrefix()

File: web/react/stores/browser_store.jsx [hunk 5]
```diff
@@ -89,6 +118,21 @@ class BrowserStoreClass {
...
+        for (var key in sessionStorage) {
+            if (key.lastIndexOf(globalPrefix + prefix, 0) === 0) {
+                var userkey = key.substring(globalPrefixiLen);
+                action(userkey, this.getGlobalItem(key));
+            }
+        }
+    }
+
+    clear() {
+        sessionStorage.clear();
```
Comment (coreyhulen) [inline web/react/stores/browser_store.jsx]: double check why localstorage isn't called

File: web/react/stores/browser_store.jsx [hunk 6]
```diff
@@ -66,19 +88,26 @@ class BrowserStoreClass {
...
 
     /**
      * Preforms the given action on each item that has the given prefix
      * Signature for action is action(key, value)
      */
-    actionOnItemsWithPrefix(prefix, action) {
+    actionOnGlobalItemsWithPrefix(prefix, action) {
         var globalPrefix = getPrefix();
         var globalPrefixiLen = globalPrefix.length;
         for (var key in localStorage) {
```
Comment (coreyhulen) [inline web/react/stores/browser_store.jsx]: this will bomb while in private viewing

File: web/web.go [hunk 7]
```diff
@@ -972,6 +1026,7 @@ func incomingWebhook(c *api.Context, w http.ResponseWriter, r *http.Request) {
 
 	// create a mock session
 	c.Session = model.Session{UserId: hook.UserId, TeamId: hook.TeamId, IsOAuth: false}
+	c.SessionTokenIndex = 0
```
Comment (coreyhulen) [inline web/web.go]: I don't need todo this

File: model/client.go (hunk 2 above)
Comment (coreyhulen) [inline model/client.go]: oops...cannot be refactored since this is in client.go

File: web/react/stores/browser_store.jsx (hunk 5 above)
Comment (coreyhulen) [inline web/react/stores/browser_store.jsx]: this is only used for logout and should be session only.

File: api/context.go [hunk 8]
```diff
@@ -489,6 +510,46 @@ func Handle404(w http.ResponseWriter, r *http.Request) {
...
+			l4g.Error("Invalid session token=" + token + ", err=" + sessionResult.Err.DetailedError)
+		} else {
+			session = sessionResult.Data.(*model.Session)
+		}
+	}
+
+	return session
+}
+
+func GetMultiSessionCookie(r *http.Request) []string {
```
Comment (jwilander) [inline api/context.go]: 0/5 but maybe call this `GetMultiSessionCookieTokens` or something similar so it's obvious what's being returned?

File: api/web_team_hub.go [hunk 9]
```diff
@@ -96,7 +96,8 @@ func ShouldSendEvent(webCon *WebConn, msg *model.Message) bool {
 		}
 	} else {
 		// Don't share a user's view events with other users
-		if msg.Action == model.ACTION_CHANNEL_VIEWED {
+		// but you still need to share it with yourself
+		if webCon.UserId != msg.UserId && msg.Action == model.ACTION_CHANNEL_VIEWED {
```
Comment (jwilander) [inline api/web_team_hub.go]: This is unnecessary since if it gets into the enclosing `else` it's not the same user

**Comment by jwilander** [comment]
+1, other than my two comments

**Pull request merged**